```{r echo=FALSE, message=FALSE, warning=FALSE}
library(dplyr)
library(devtools)
library(GenomicRanges)
library(intSiteRetriever)
library(hiAnnotator)
library(BSgenome)
library(pipeUtils)

source_url('https://raw.githubusercontent.com/BushmanLab/genomicHeatmapMaker/master/genomicHeatmapMaker.R')
source_url('https://raw.githubusercontent.com/BushmanLab/EpigeneticHeatmapMaker/master/epigeneticHeatMapMaker.R')
source_url('https://raw.githubusercontent.com/BushmanLab/genomicHeatmapMaker/master/utils.R')
```

# Sites Observed in Cell Types

Lets load reference genome and sequence lenghts of chromosomes:
```{r}
referenceGenome <- "hg18"
reference <- get_reference_genome(referenceGenome)
```

Cell types, cell counts and genomic location data:
```{r read_fdr}
cellType_location <- read.csv("fdr-empir.csv")
columns <- c("Neutrophil", "Bcell", "Tcell", "Monocytes", "NKcell", "location")
cell_location <- select(cellType_location, one_of(columns)) 
str(cell_location)
```

For each cell type(N cells > 0) we need to get genomic locations of sites:
```{r split_based_on_cell_type}
cell_types <- c("Neutrophil", "Bcell", "Tcell", "Monocytes", "NKcell")
cell_types
sites_for_cell_type <- lapply(cell_types, function(cell_type) {
    cell_type_present <- filter_(cell_location, paste(cell_type, " > 0"))
    data_frame(location=cell_type_present$location, sampleName=cell_type) 
})
sites <- do.call(rbind, sites_for_cell_type)
str(sites)
table(sites$sampleName)
```

Add metadata for heatmaps to connect conrols and observed sites:
```{r}
sites$type <- "insertion" # real integration event
sites$siteID <- seq(1, nrow(sites)) + 12345
```
We are adding random constant as a seed for control(MRCs) generation.

Now we need to create several(N_MRCs) matched random controls:
```{r generate_controls}
create_MRC <- function(sites, reference) {
  N_MRCs <- 3
  siteIDs <- sites$siteID
  mrcs <- get_random_positions(siteIDs, reference, number_of_positions = N_MRCs)
  mrcs$sampleName <- rep(sites$sampleName, each=N_MRCs)
  mrcs$type <- "match"
  mrcs$location <- paste0(mrcs$chr, ":", mrcs$position, "-", mrcs$position, ":", mrcs$strand)
  mrcs[c("location", "sampleName", "type", "siteID")]
}
mrcs <- create_MRC(sites, reference)
```

Now we have real sites and controls and required metadata and can convert it to
GRanges object. 'SiteID' to link sites and controls, 'type' to distinguish them
and 'sampeName' for final labeling on plot:
```{r merge_sites_controls}
cat_sites_MRC <- function (sites, mrcs, reference) {
  sites_mrcs <- rbind(sites, mrcs)
  sites_meta <- sites_mrcs[c("sampleName", "type", "siteID")]
  sites_mrcs <- GRanges(sites_mrcs$location, seqinfo=seqinfo(reference))
  mcols(sites_mrcs) <- sites_meta
  sites_mrcs$chr <- seqnames(sites_mrcs)
  sites_mrcs$position <- start(sites_mrcs)
  sites_mrcs
}
sites_mrcs <- cat_sites_MRC(sites, mrcs, reference)
```

# Genomic Heatmap

We now can pull genomic features from UCSC and use `pipeUtils` package to 
find if distribution of sites are different for real sites and controls:

```{r genomic_heatmap}
sites_to_heatmap(sites_mrcs, referenceGenome, "genomic_heatmap")
```

Description of the features and ROC  interpretation can be found at 
[RosettaStone](https://github.com/BushmanLab/genomicHeatmapMaker/blob/master/genomicFeaturesRosettaStone.pdf).

# Epigenetic Heatmap

Epigenetic features are stored locally:
```{r}
annotation_path <- file.path("/media/THING1/dryga/Epigenetic/", referenceGenome)
histoneorder <- epigenetic_features()
```

Among all of the epigenetic features we are going to use CD133+ only:
```{r}
histoneorder <- grep('CD133', histoneorder, value=TRUE)
histoneorder
```

Now we can generate epigenetic heatmap:
```{r epi_heatmap}
sites_to_epigenetic_heatmap(sites_mrcs, referenceGenome, "epigenetic_heatmap", 
    annotation_path, histoneorder)
```

# Sites Seen Predominantly in Single Cell Type

Here we perform the same analysis but for sites seen only(more accurately - mostly)
in single cell type. The formal definition is: sites that have 3 or more cells 
counted in only single cell type.

Lets calculate presence/absence for all cell types:
```{r}
N_CUT_OFF <- 3
cellPresent_location <- cell_location
res <- lapply(cell_types, function(cell_type) {
    cellPresent_location[ , cell_type] <<- cellPresent_location[ , cell_type] >= N_CUT_OFF
})
```

Now we can find cases when only one cell type present:
```{r}
cellPresent_location$is_singleton <- 0
res <- lapply(cell_types, function(cell_type) {
    cellPresent_location$is_singleton <<- 
        cellPresent_location$is_singleton + as.integer(cellPresent_location[ , cell_type])
    # using type conversion: T to 1, F to 0
})
cellPresent_location <- mutate(cellPresent_location, is_singleton = (is_singleton == 1))
```

Most of the sites observed in multiple cell types:
```{r}
table(cellPresent_location$is_singleton)
```

Lets get sites that only observed only in one cell type, that would require removal
of some cell types:
```{r}
cellPresent_location <- filter(cellPresent_location, is_singleton == TRUE)
sites_count_cell_type <- sapply(cell_types, function(cell_type) {
    sum(cellPresent_location[ , cell_type])
})
sites_count_cell_type
N_SITES_CUT_OFF <- 3
cell_types <- names(sites_count_cell_type)[sites_count_cell_type > N_SITES_CUT_OFF]
cell_types
```

For cell types with enough sites:
```{r}
singleton_sites_for_cell_type <- lapply(cell_types, function(cell_type) {
    cell_type_present <- filter_(cellPresent_location, paste(cell_type, " == TRUE"))
    data_frame(location=cell_type_present$location, sampleName=cell_type) 
})
singleton_sites <- do.call(rbind, singleton_sites_for_cell_type)
table(singleton_sites$sampleName)
```

Add metadata for heatmaps to connect conrols and observed sites:
```{r}
singleton_sites$type <- "insertion" # real integration event
singleton_sites$siteID <- seq(1, nrow(sites)) + 3141592
```

Now we can add controls and create final Granges object:
```{r singleton_sites_mrcs}
singleton_sites_mrcs <- cat_sites_MRC(singleton_sites, 
    create_MRC(singleton_sites, reference), reference)
```

Genomic heatmaps:
```{r singleton_genomic_heatmap}
sites_to_heatmap(singleton_sites_mrcs, referenceGenome, "singleton_genomic_heatmap")
```

Epigentic heatmaps:
```{r singleton_epigentic_heatmap}
sites_to_epigenetic_heatmap(singleton_sites_mrcs, referenceGenome, 
    "singleton_epigenetic_heatmap", annotation_path, histoneorder)
```
